File: event/event_dispatcher.py
import threading
import logging
from typing import Dict, Set
from event.event import Event
from event.event_type import EventType
from event.event_consumer import EventConsumer
from event.event_queue import EventQueue

class EventDispatcher:
    def __init__(self):
        self.event_queue = EventQueue()
        self.consumer_map: Dict[EventType, Set[EventConsumer]] = {}
        self.thread = None
        self.logger = logging.getLogger(__name__)
    
    def start(self):
        self.thread = threading.Thread(target=self.run)
        self.thread.daemon = True
        self.thread.name = self.__class__.__name__
        self.thread.start()
    
    def add_subscription(self, event_type: EventType, consumer: EventConsumer):
        if event_type not in self.consumer_map:
            self.consumer_map[event_type] = set()
        self.consumer_map[event_type].add(consumer)
    
    def dispatch(self, event: Event):
        self.event_queue.put(event)
    
    def run(self):
        while True:
            try:
                event = self.event_queue.get()
                consumers = self.consumer_map.get(event.get_type())
                if consumers:
                    for consumer in consumers:
                        consumer.receive_event(event)
                self.event_queue.task_done()
            except Exception as e:
                self.logger.error(f"Error processing event: {e}")

File: lapras-middleware/src/main/java/kr/ac/kaist/cdsn/lapras/event/EventDispatcher.java
package kr.ac.kaist.cdsn.lapras.event;

import kr.ac.kaist.cdsn.lapras.Component;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArraySet;

/**
 * Created by Daekeun Lee on 2016-11-11.
 */
public class EventDispatcher implements Runnable {
    private Thread thread;
    private EventQueue eventQueue;
    private Map<EventType, Set<EventConsumer>> consumerMap;

    public EventDispatcher() {
        this.eventQueue = new EventQueue();
        this.consumerMap = new ConcurrentHashMap<>();
    }

    public void start() {
        this.thread = new Thread(this);
        this.thread.setName(EventDispatcher.class.getSimpleName());
        this.thread.setDaemon(true);

        this.thread.start();
    }

    public void addSubscription(EventType eventType, Component component) {
        Set<EventConsumer> consumers = this.consumerMap.getOrDefault(eventType, new CopyOnWriteArraySet<>());
        consumers.add(component);
        this.consumerMap.put(eventType, consumers);
    }

    public void dispatch(Event event) throws InterruptedException {
        this.eventQueue.put(event);
    }

    public void run() {
        while(true) {
            Event event;
            try {
                event = eventQueue.take();
            } catch (InterruptedException e) {
                return;
            }
            Set<EventConsumer> consumers = this.consumerMap.get(event.getType());
            if(consumers == null) continue;
            for(EventConsumer consumer : consumers) {
                consumer.receiveEvent(event);
            }
        }
    }
}
